File: dsa_python/LinkedList.py
from Node import Node

class LinkedList:
    def __init__(self, data):
        self._head = Node(data)  # Head Node
        self._count = 1
        self_tail = None

    def insert_as_head(self, data):
        new_node = Node(data)
        new_node.set_next(self._head)
        self._head = new_node
        self._count += 1

    def insert_after(self,node, data):
        new_node = Node(data)
        new_node.set_next(node.get_next())
        node.set_next(new_node)

    def delete_head(self):
        if self._head is None:
            raise Exception("Empty Stack")

        # IF there is no next node, return 'None' Data
        if self._head.get_data() is None:
            return None
        else:
            popped_element = self._head.get_data()
            self._head = self._head.get_next()
            return popped_element

    def get_data(self):
        return self._head.get_data()

    def set_head_data(data):
        self._head.set_data(data)
        
    def get_head_node(self):
        return self._head

    def remove_node(self,data):
        #Like Pointer and pre ptr
        pre_tmp_node = self.get_head_node()
        pre_tmp_data = self.get_head_node().get_data()
        tmp_node = self.get_head_node().get_next()
        tmp_data = tmp_node.get_data()
        
        while(True):
            if(pre_tmp_node == None): # TODO
                break
            if(tmp_data == data):
                pre_tmp_node.set_next(tmp_node.get_next())
            else:
                # Increment Node
                pre_tmp_node = pre_tmp_node.get_next()
                tmp_node = tmp_node.get_next()
                # Set Data
                tmp_data = tmp_node.get_data()
                pre_tmp_data = pre_tmp_data.get_data()

    def get_count():
        return self._count

File: ds_cpp/linked_list.cpp
#include "linked_list.h"
#include <iostream>

template<typename type>
linked_list<type>::linked_list(type data) {
    _head = new Node<type>(data);
    _tail = _head;
    if(data == NULL) _count = -1;
    else _count = 1;
}

template<typename type>
void linked_list<type>::push(type data) {
//    if(_head == nullptr){
    if(_count == -1){
        _head = new Node<type>(data); // As list is empty, there is no head
        _tail = _head;
        _count = 1; // As in empty linked list, _count = -1
    }
    else {
    _count++;
    _tail->set_next(new Node<type>(data));
    _tail = _tail->get_next();
    }
}

template<typename type>
void linked_list<type>::insert_as_head(Node<type> *node) {
    node->set_next(_head);
    _head = node;
    if (_head == nullptr) _count = 1;
    else _count++;
}



template<typename type>
type linked_list<type>::delete_head() {
    if(_head->get_next() == nullptr )
//    if(_head == nullptr )
    {
        if(_count == -1) throw std::runtime_error("List is empty");
        _count = -1;
        auto tmp = _head->get_data();
        delete _head;
        _head,_tail = nullptr;
        return tmp; // Method Ends here
    }
    else{
        auto tmp_ptr = _head;
        auto tmp_data = _head->get_data();
        _head = _head->get_next();
        delete tmp_ptr;
        _count--;
        return tmp_data;
    }
}

template<typename type>
void linked_list<type>::insert_after(Node<type> *old_node, Node<type> *new_node) {
    new_node->set_next(old_node->get_next());
    old_node->set_next(new_node);
}

template <typename type>
void linked_list<type>::insert_as_tail(type data)
{
    insert_after(_tail, new Node<type>);
}

template<typename type>
type linked_list<type>::get_head_data() const {
    if(_count == -1) throw std::runtime_error("List is empty");
    return _head->get_data();
}

template<typename type>
void linked_list<type>::insert_as_head(type data) {
    auto node =  new Node<type>(data);
    if(_head == nullptr) // Linked list is empty
    {
        node->set_next(nullptr); // TODO
        _tail = node;
    }
    else
        node->set_next(_head);
    _head = node;
    if (_head == nullptr) _count = 1;
    else _count++;
}

template<typename type>
void linked_list<type>::push(Node<type> *node) {
//    if(_head == nullptr){
    if(_count == -1){
        _head = node; // As list is empty, there is no head
        _tail = _head;
        _count = 1; // As in empty linked list, _count = -1
    }
    else {
        _count++;
        _tail->set_next(node);
        _tail = _tail->get_next();
    }
}

